File: POSmaisgestao/app/src/main/java/com/pos_mais_gestao/util/PizzaMeioAMeio.java
package com.pos_mais_gestao.util;

import com.pos_mais_gestao.domain.Produto;
import java.math.BigDecimal;

/**
 * Pizza meio a meio: dois produtos com {@code espizza=1}.
 * Preço = o maior entre as metades. Quantidade = 1 pizza.
 * NFC-e/estoque usam o produto de maior preço como base.
 */
public final class PizzaMeioAMeio {
    public static final int DESCRICAO_NFCE_MAX = 120;

    private PizzaMeioAMeio() {}

    public static String descricao(Produto primeiro, Produto segundo) {
        String texto =
                "Pizza meio a meio: " + primeiro.getDescricao() + " / " + segundo.getDescricao();
        if (texto.length() <= DESCRICAO_NFCE_MAX) {
            return texto;
        }
        return texto.substring(0, DESCRICAO_NFCE_MAX);
    }

    public static BigDecimal preco(Produto primeiro, Produto segundo) {
        return primeiro.getPreco().max(segundo.getPreco());
    }

    /** Produto fiscal/estoque/impressão: o de maior preço (empate: o primeiro). */
    public static Produto principal(Produto primeiro, Produto segundo) {
        if (primeiro.getPreco().compareTo(segundo.getPreco()) >= 0) {
            return primeiro;
        }
        return segundo;
    }

    public static Produto outro(Produto primeiro, Produto segundo) {
        Produto base = principal(primeiro, segundo);
        return base.getId().equals(primeiro.getId()) ? segundo : primeiro;
    }
}

File: pdv/src/lib/pizza-meio-a-meio.ts
/** Limite de xProd na NFC-e. */
export const DESCRICAO_NFCE_MAX = 120;

export type ProdutoPizza = {
	id: string;
	descricao: string;
	preco: number;
	espizza?: number | boolean | null;
};

export function produtoEhPizza(produto: {
	espizza?: number | boolean | string | null;
}): boolean {
	return Number(produto.espizza) === 1 || produto.espizza === true;
}

export function montarDescricaoPizzaMeioAMeio(
	primeiro: string,
	segundo: string,
): string {
	const texto = `Pizza meio a meio: ${primeiro.trim()} / ${segundo.trim()}`;
	if (texto.length <= DESCRICAO_NFCE_MAX) {
		return texto;
	}
	return texto.slice(0, DESCRICAO_NFCE_MAX);
}

/**
 * Monta 1 pizza meio a meio a partir de dois produtos pizza.
 * Preço = maior das metades. idproduto fiscal = o de maior preço (empate: o primeiro).
 * Impressão de produção segue o grupo gourmet desse produto-base.
 */
export function montarItemPizzaMeioAMeio(
	primeiro: ProdutoPizza,
	segundo: ProdutoPizza,
): {
	idproduto: string;
	idprodutomeio: string;
	descricao: string;
	quantidade: number;
	precounitario: number;
	precototal: number;
} {
	const precoPrimeiro = Number(primeiro.preco) || 0;
	const precoSegundo = Number(segundo.preco) || 0;
	const principal = precoPrimeiro >= precoSegundo ? primeiro : segundo;
	const outro = principal.id === primeiro.id ? segundo : primeiro;
	const precounitario = Math.max(precoPrimeiro, precoSegundo);
	return {
		idproduto: principal.id,
		idprodutomeio: outro.id,
		descricao: montarDescricaoPizzaMeioAMeio(
			primeiro.descricao,
			segundo.descricao,
		),
		quantidade: 1,
		precounitario,
		precototal: precounitario,
	};
}
